pkg/translate/translate.go

```diff
@@ -0,0 +1,566 @@
+					return err
+				}
+			}
+			if c.namespace != "" {
+				if err := setTree(root, util.PathFromString(c.namespace), name.Namespace(string(fn), cn, ii)); err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// insertLeaf inserts a leaf with value into root at path, which is first mapped using t.APIMapping.
+func (t *Translator) insertLeaf(root map[string]interface{}, path util.Path, value reflect.Value) (errs util.Errors) {
+	// Must be a scalar leaf. See if we have a mapping.
+	valuesPath, m := getValuesPathMapping(t.APIMapping, path)
+	var v interface{}
+	if value.Kind() == reflect.Ptr {
+		v = value.Elem().Interface()
+	} else {
+		v = value.Interface()
+	}
+	switch {
+	case m == nil:
+		break
+	case m.translationFunc == nil:
+		// Use default translation which just maps to a different part of the tree.
+		errs = util.AppendErr(errs, defaultTranslationFunc(m, root, valuesPath, v))
+	default:
+		// Use a custom translation function.
+		errs = util.AppendErr(errs, m.translationFunc(m, root, valuesPath, v))
+	}
+	return errs
+}
+
+// getValuesPathMapping tries to map path against the passed in mappings with a longest prefix match. If a matching prefix
+// is found, it returns the translated YAML path and the corresponding translation.
+// e.g. for mapping "a.b"  -> "1.2", the input path "a.b.c.d" would yield "1.2.c.d".
+func getValuesPathMapping(mappings map[string]*Translation, path util.Path) (string, *Translation) {
+	p := path
+	var m *Translation
+	for ; len(p) > 0; p = p[0 : len(p)-1] {
+		m = mappings[p.String()]
+		if m != nil {
+			break
+		}
+	}
+	if m == nil {
+		return "", nil
+	}
+
+	if m.outPath == "" {
+		return "", m
+	}
+
+	out := m.outPath + "." + path[len(p):].String()
+	dbgPrint("translating %s to %s", path, out)
+	return out, m
+}
+
+// featureComponentString renders a template of the form <path>{{.FeatureName}}<path>{{.ComponentName}}<path> with
+// the supplied parameters.
+func featureComponentString(tmpl string, featureName name.FeatureName, componentName name.ComponentName) string {
+	type Temp struct {
+		FeatureName   name.FeatureName
+		ComponentName name.ComponentName
+	}
+	ts := Temp{
+		FeatureName:   featureName,
+		ComponentName: componentName,
+	}
+	t, err := template.New("").Parse(tmpl)
+	if err != nil {
+		log.Error(err.Error())
+		return err.Error()
+	}
+	buf := new(bytes.Buffer)
+	err = t.Execute(buf, ts)
+	if err != nil {
+		log.Error(err.Error())
+		return err.Error()
+	}
+	return buf.String()
+}
+
+// resourceContainerString renders a template of the form <path>{{.ResourceName}}<path>{{.ContainerName}}<path> with
+// the supplied parameters.
+func (t *Translator) resourceContainerString(tmpl string, componentName name.ComponentName) string {
+	ts := struct {
+		ResourceName  string
+		ContainerName string
+	}{
+		ResourceName:  t.ComponentMaps[componentName].ResourceName,
+		ContainerName: t.ComponentMaps[componentName].ContainerName,
+	}
+	tm, err := template.New("").Parse(tmpl)
```

Comment:
Can extract the template execution part to a common method, so for each rendering method just need to create a template struct and call this common method

Reply:
Left as a TODO